File: src/main/java/nl/dannyjelsma/soundboardbot/music/PlaylistManager.java
package nl.dannyjelsma.soundboardbot.music;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import nl.dannyjelsma.soundboardbot.audio.SoundManager;
import nl.dannyjelsma.soundboardbot.result.PlaylistCreationResult;
import nl.dannyjelsma.soundboardbot.result.PlaylistLoadResult;
import nl.dannyjelsma.soundboardbot.result.PlaylistResult;

import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

public class PlaylistManager {

    public List<File> getPlaylists() {
        File[] files = new File("./playlists/").listFiles();
        List<File> playlists = new ArrayList<>();

        if (files != null) {
            for (File file : files) {
                if (file.getName().endsWith(".json")) {
                    playlists.add(file);
                }
            }
        }

        return playlists;
    }

    public PlaylistCreationResult createPlaylist(String name) {
        ObjectMapper mapper = new ObjectMapper();
        Playlist playlist = new Playlist();
        File newPlaylist = new File("./playlists/", name + ".json");

        if (newPlaylist.exists()) {
            return new PlaylistCreationResult(false, "This playlist already exists!", newPlaylist);
        }

        if (!newPlaylist.getParentFile().exists()) {
            newPlaylist.getParentFile().mkdir();
        }

        try {
            newPlaylist.createNewFile();
            mapper.writeValue(newPlaylist, playlist);
            return new PlaylistCreationResult(true, null, newPlaylist);
        } catch (Exception e) {
            return new PlaylistCreationResult(false, e.getMessage(), newPlaylist);
        }
    }

    public PlaylistResult deletePlaylist(String name) {
        File playlist = new File("./playlists/", name + ".json");
        File dstFile = new File("./playlists/", name + ".json.disabled");

        if (!playlist.exists()) {
            return new PlaylistResult(false, "That playlist doesn't exist!");
        }

        try {
            Files.move(playlist.toPath(), dstFile.toPath());
        } catch (IOException e) {
            playlist.delete();
            e.printStackTrace();
        }

        return new PlaylistResult(true, null);
    }

    public PlaylistResult addToPlaylist(String playlistName, String url) {
        PlaylistLoadResult loadResult = loadPlaylist(playlistName);

        if (!loadResult.isSuccess()) {
            return new PlaylistResult(false, "Could not load playlist: " + loadResult.getMessage());
        }

        Playlist playlist = loadResult.getPlaylist();
        String query = url;

        try {
            new URL(query);
        } catch (MalformedURLException e) {
            query = "ytsearch:" + query;
        }

        try {
            SoundManager manager = SoundManager.getInstance();
            final boolean[] success = {false};
            String finalQuery = query;

            if (manager.isInCache(finalQuery)) {
                AudioTrack track = manager.getFromCache(finalQuery);
                Song song = new Song(track.getInfo().title, track.getInfo().uri);

                playlist.addSong(song);
                savePlaylist(playlist);
                success[0] = true;
            } else {
                manager.getAudioPlayerManager().loadItemOrdered(manager, query, new AudioLoadResultHandler() {
                    @Override
                    public void trackLoaded(AudioTrack track) {
                        Song song = new Song(track.getInfo().title, track.getInfo().uri);

                        manager.addToCache(finalQuery, track);
                        playlist.addSong(song);
                        savePlaylist(playlist);
                        success[0] = true;
                    }

                    @Override
                    public void playlistLoaded(AudioPlaylist audioPlaylist) {
                        if (finalQuery.contains("/playlist")) {
                            for (AudioTrack track : audioPlaylist.getTracks()) {
                                Song song = new Song(track.getInfo().title, track.getInfo().uri);

                                playlist.addSong(song);
                            }
                        } else {
                            AudioTrack track = audioPlaylist.getTracks().get(0);
                            Song song = new Song(track.getInfo().title, track.getInfo().uri);

                            manager.addToCache(finalQuery, track);
                            playlist.addSong(song);
                        }

                        savePlaylist(playlist);
                        success[0] = true;
                    }

                    @Override
                    public void noMatches() {
                    }

                    @Override
                    public void loadFailed(FriendlyException exception) {
                    }
                }).get();
            }

            return new PlaylistResult(success[0], "No matches for this song.");
        } catch (Exception e) {
            return new PlaylistResult(false, e.getMessage());
        }
    }

    public PlaylistResult removeFromPlaylist(String playlistName, String url) {
        PlaylistLoadResult loadResult = loadPlaylist(playlistName);

        if (!loadResult.isSuccess()) {
            return new PlaylistResult(false, "Could not load playlist: " + loadResult.getMessage());
        }

        Playlist playlist = loadResult.getPlaylist();
        List<Song> songs = playlist.getSongs();

        for (Song song : songs) {
            if (song.getUrl().equals(url)) {
                playlist.removeSong(song);
                savePlaylist(playlist);
                return new PlaylistResult(true, null);
            }
        }

        return new PlaylistResult(false, "This URL is not in this playlist.");
    }

    public PlaylistLoadResult loadPlaylist(String playlistName) {
        ObjectMapper mapper = new ObjectMapper();

        try {
            File playlistFile = new File("./playlists/", playlistName + ".json");
            Playlist playlist = mapper.readValue(playlistFile, Playlist.class);

            playlist.setPlaylistFile(playlistFile);
            return new PlaylistLoadResult(true, null, playlist);
        } catch (Exception e) {
            return new PlaylistLoadResult(false, e.getMessage(), null);
        }
    }

    public PlaylistResult savePlaylist(Playlist playlist) {
        ObjectMapper mapper = new ObjectMapper();

        try {
            mapper.writeValue(playlist.getPlaylistFile(), playlist);

            return new PlaylistResult(true, null);
        } catch (Exception e) {
            return new PlaylistResult(false, e.getMessage());
        }
    }

    public PlaylistResult removeFromPlaylistByTitle(String playList, String title) {
        PlaylistLoadResult loadResult = loadPlaylist(playList);

        if (!loadResult.isSuccess()) {
            return new PlaylistResult(false, "Could not load playlist: " + loadResult.getMessage());
        }

        Playlist playlist = loadResult.getPlaylist();
        List<Song> songs = playlist.getSongs();

        for (Song song : songs) {
            if (song.getTitle().equalsIgnoreCase(title)) {
                playlist.removeSong(song);
                savePlaylist(playlist);
                return new PlaylistResult(true, null);
            }
        }

        return new PlaylistResult(false, "This URL is not in this playlist.");
    }
}

File: src/main/java/nl/dannyjelsma/soundboardbot/music/PlaylistManager.kt
package nl.dannyjelsma.soundboardbot.music

import com.fasterxml.jackson.databind.ObjectMapper
import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import nl.dannyjelsma.soundboardbot.audio.SoundManager
import nl.dannyjelsma.soundboardbot.result.PlaylistCreationResult
import nl.dannyjelsma.soundboardbot.result.PlaylistLoadResult
import nl.dannyjelsma.soundboardbot.result.PlaylistResult
import java.io.File
import java.io.IOException
import java.net.MalformedURLException
import java.net.URL
import java.nio.file.Files

class PlaylistManager {
    val playlists: List<File>
        get() {
            val files = File("./playlists/").listFiles()
            val playlists: MutableList<File> = ArrayList()
            if (files != null) {
                for (file in files) {
                    if (file.name.endsWith(".json")) {
                        playlists.add(file)
                    }
                }
            }
            return playlists
        }

    fun createPlaylist(name: String): PlaylistCreationResult {
        val mapper = ObjectMapper()
        val playlist = Playlist()
        val newPlaylist = File("./playlists/", "$name.json")
        if (newPlaylist.exists()) {
            return PlaylistCreationResult(false, "This playlist already exists!")
        }

        if (!newPlaylist.parentFile.exists()) {
            newPlaylist.parentFile.mkdir()
        }

        return try {
            newPlaylist.createNewFile()
            mapper.writeValue(newPlaylist, playlist)
            PlaylistCreationResult(true, null)
        } catch (e: Exception) {
            PlaylistCreationResult(false, e.message)
        }
    }

    fun deletePlaylist(name: String): PlaylistResult {
        val playlist = File("./playlists/", "$name.json")
        val dstFile = File("./playlists/", "$name.json.disabled")
        if (!playlist.exists()) {
            return PlaylistResult(false, "That playlist doesn't exist!")
        }

        try {
            Files.move(playlist.toPath(), dstFile.toPath())
        } catch (e: IOException) {
            playlist.delete()
            e.printStackTrace()
        }

        return PlaylistResult(true, null)
    }

    fun addToPlaylist(playlistName: String, url: String): PlaylistResult {
        val loadResult = loadPlaylist(playlistName)
        if (!loadResult.isSuccess) {
            return PlaylistResult(false, "Could not load playlist: " + loadResult.message)
        }

        val playlist = loadResult.playlist
        if (playlist != null) {
            var query = url

            try {
                URL(query)
            } catch (e: MalformedURLException) {
                query = "ytsearch:$query"
            }

            return try {
                val success = booleanArrayOf(false)
                val audioTrack: AudioTrack? = SoundManager.getFromCache(query)
                if (audioTrack != null) {
                    val song = Song(audioTrack.info.title, audioTrack.info.uri)

                    playlist.addSong(song)
                    savePlaylist(playlist)
                    success[0] = true
                } else {
                    SoundManager.audioPlayerManager.loadItemOrdered(SoundManager,
                        query,
                        object : AudioLoadResultHandler {
                            override fun trackLoaded(track: AudioTrack) {
                                val song = Song(track.info.title, track.info.uri)

                                SoundManager.addToCache(query, track)
                                playlist.addSong(song)
                                savePlaylist(playlist)
                                success[0] = true
                            }

                            override fun playlistLoaded(audioPlaylist: AudioPlaylist) {
                                if (query.contains("/playlist")) {
                                    for (track in audioPlaylist.tracks) {
                                        val song = Song(track.info.title, track.info.uri)
                                        playlist.addSong(song)
                                    }
                                } else {
                                    val track = audioPlaylist.tracks[0]
                                    val song = Song(track.info.title, track.info.uri)
                                    SoundManager.addToCache(query, track)
                                    playlist.addSong(song)
                                }

                                savePlaylist(playlist)
                                success[0] = true
                            }

                            override fun noMatches() {}
                            override fun loadFailed(exception: FriendlyException) {}
                        }).get()
                }

                PlaylistResult(success[0], "No matches for this song.")
            } catch (e: Exception) {
                PlaylistResult(false, e.message)
            }
        } else {
            return PlaylistResult(false, loadResult.message)
        }
    }

    fun removeFromPlaylist(playlistName: String, url: String): PlaylistResult {
        val loadResult = loadPlaylist(playlistName)
        if (!loadResult.isSuccess) {
            return PlaylistResult(false, "Could not load playlist: " + loadResult.message)
        }

        val playlist = loadResult.playlist

        if (playlist != null) {
            val songs = playlist.songs
            for (song in songs) {
                if (song.url == url) {
                    playlist.removeSong(song)
                    savePlaylist(playlist)
                    return PlaylistResult(true, null)
                }
            }
        } else {
            return PlaylistResult(false, loadResult.message)
        }

        return PlaylistResult(false, "This URL is not in this playlist.")
    }

    fun loadPlaylist(playlistName: String): PlaylistLoadResult {
        val mapper = ObjectMapper()

        return try {
            val playlistFile = File("./playlists/", "$playlistName.json")
            val playlist = mapper.readValue(playlistFile, Playlist::class.java)
            playlist.playlistFile = playlistFile
            PlaylistLoadResult(true, null, playlist)
        } catch (e: Exception) {
            PlaylistLoadResult(false, e.message, null)
        }
    }

    fun savePlaylist(playlist: Playlist): PlaylistResult {
        val mapper = ObjectMapper()

        return try {
            mapper.writeValue(playlist.playlistFile, playlist)
            PlaylistResult(true, null)
        } catch (e: Exception) {
            PlaylistResult(false, e.message)
        }
    }

    fun removeFromPlaylistByTitle(playList: String, title: String): PlaylistResult {
        val loadResult = loadPlaylist(playList)
        if (!loadResult.isSuccess) {
            return PlaylistResult(false, "Could not load playlist: " + loadResult.message)
        }

        val playlist = loadResult.playlist
        if (playlist != null) {
            val songs = playlist.songs
            for (song in songs) {
                if (song.title.equals(title, ignoreCase = true)) {
                    playlist.removeSong(song)
                    savePlaylist(playlist)
                    return PlaylistResult(true, null)
                }
            }

            return PlaylistResult(false, "This URL is not in this playlist.")
        } else {
            return PlaylistResult(false, loadResult.message)
        }
    }
}